File: intervals/cppRealizations/meetingRoomsII.cpp
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;
 

class Interval {
public:
    int start, end;
    Interval(int start, int end) {
        this->start = start;
        this->end = end;
    }
};


class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
        vector<int> starts, ends;
        for (Interval interval : intervals) {
            starts.push_back(interval.start);
            ends.push_back(interval.end);
        }
        sort(starts.begin(), starts.end());
        sort(ends.begin(), ends.end());
        int res = 0, count = 0;
        int s = 0, e = 0;
        while(s < intervals.size()){
            if(starts[s] < ends[e]){
                count++;
                s++;
            }else{
                count--;
                e++;
            }
            res = max(res, count);
        }
        return res;
    }
};



int main(){
    vector<Interval> intervals = {Interval(0, 40), Interval(5, 10), Interval(15, 20)};
    // vector<Interval> intervals = {{0, 10}, {1, 3}, {2, 6}, {5, 8}, {7, 12}, {11, 15}, {13, 18}, {16, 20}, {19, 25}, {24, 30}};
    cout << "Output:" << endl;
    cout << Solution().minMeetingRooms(intervals) << endl;
    return 0;
}

File: intervals/meetingRoomsII.py
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end


class Solution:
    def minMeetingRooms(self, intervals: list[Interval]) -> bool:
        start = sorted([x.start for x in intervals])
        end = sorted([x.end for x in intervals])

        res, count = 0, 0
        s, e = 0, 0
        while s < len(intervals):
            if start[s] < end[e]:
                count += 1
                s += 1
            else:
                count -= 1
                e += 1
            res = max(res, count)
        return res


if __name__ == "__main__":
    intervals = [Interval(0, 40), Interval(5, 10), Interval(15, 20)]
    # intervals = [Interval(0, 10), Interval(1, 3), Interval(2, 6), Interval(5, 8),Interval(7, 12), Interval(11, 15), Interval(13, 18), Interval(16, 20),Interval(19, 25), Interval(24, 30)]

    print("Solution:")
    print(Solution().minMeetingRooms(intervals))